### ISSUE TITLE
Oddness with Salford postcodes

### ISSUE THREAD

**Issue by symroe** [DemocracyClub/UK-Polling-Stations#588]
For example, this is clearly wrong:

https://wheredoivote.co.uk/postcode/M65LZ/

**Comment by symroe** [comment]
Hmm, well, what I suggested is a good idea, but it looks like the morph point data is correct here, so is something else going on? I'll rename this issue

**Comment by chris48s** [comment]
My hunch was correct. Here's what is going on:

Mapit is putting the point for M6 5LZ in the correct place (in Salford):
http://mapit.democracyclub.org.uk/postcode/M65LZ.html
https://mapit.mysociety.org/postcode/M65LZ.html
and we are (correctly) looking up the district that point is in and assigning it the relevant polling station.

Then when we generate the directions line, we are calling

https://maps.googleapis.com/maps/api/directions/json?mode=walking&units=imperial&key=&origin=M65LZ&destination=53.48364772,-2.29406961

Google's directions API is (incorrectly) geocoding the source point (M65LZ) to "M6, Knutsford WA16, UK" and using that as the source point for the directions line - and then we're displaying that line (note that the 'home' point is the one that is wrong - not the polling station). This is particularly odd because the directions _link_ on that page https://www.google.com/maps/dir/M65LZ/53.48364772,-2.29406961 does pick up the source point correctly.

I think the solution to this is to use the _point_ we've already derived for the postcode centroid as the origin argument when generating the directions instead of the postcode so this type of ambiguous situation can not arise. I've not tested it but since we're passing the destination as a co-ordinate, I assume the origin can be too. I will have a look at this.

With respect to your other comment about stations outside the local auth, for a shapes and points importer, if there are any stations which don't fall inside a district this will be reported in the data quality report as 'stations in 0 districts' and I do look at this when I am checking the data: If a number other than 0 is reported, I have a look at the data and see if it seems reasonable (i.e: it is _just_ outside the local auth and serves a district on the edge) or looks like a mistake (i.e: it is _miles_ off). In that case I feed back to Joe to query it. We don't have this check for ResidentialAddress importers although often these don't have points for the polling stations and we rely on postcode centroid from mapit anyway. It might be worth adding a check for this.

Note that it is sometimes valid for a polling station to be located just outside the local auth. I do occasionally see this, but I can't find an example to hand. I don't really have a feel for what a reasonable margin would be so I think reporting as a possible issue to review in the data quality report is the correct approach.

**Comment by symroe** [comment]
I've made this change in commit ece346aefcb336f0da29f148e99eee4853da7085 to my feature branch that I'm using to test deployment. Seems to work. We'll have to talk about how/when we merge that branch in to master.

**Comment by chris48s** [comment]
I've submitted a PR for it at #602 which includes updates for the acceptance tests, which have [failed](https://travis-ci.org/DemocracyClub/UK-Polling-Stations/builds/213759220) on your branch, so we can merge that independently.

**Comment by chris48s** [comment]
Hmm.. we seem to be getting some issues with this popping up in the logs. e.g:

https://wheredoivote.co.uk/address/10093262074/

throwing
```
  File "/var/www/polling_stations/code/polling_stations/apps/data_finder/helpers.py", line 202, in get_google_route
    start_y=start.y,
AttributeError: 'NoneType' object has no attribute 'y'
```

**Comment by chris48s** [comment]
actually, that was quite an obvious bug: fix proposed in 8e236c1
